fix: Fixed tooltip for Mobile Area Charts

noahonyejese opened the pull request:
@bprusinowski please check if this is the correct approach, if so I will continue implementing this for all charts and for tooltips that are larger the arrow will have to be more dynamic of course.

This PR: 
- Fixes Mobile Tooltip for all Chart except (Pie Charts, Scatterplot Charts, Map Charts)
- The exceptions still need design criteria, @KerstinFaye 
- This PR also uncovered a new Issue [#110](https://github.com/visualize-admin/visualization-tool-private/issues/110)
- This PR is the start to fix Issue [#94](https://github.com/visualize-admin/visualization-tool-private/issues/94)

This PR will resolve #94

---
### How to Test
1. Go to this [Visit Preview](https://vercel.live/open-feedback/visualization-tool-git-fix-mobile-tooltip-positioning-ixt1.vercel.app?via=pr-comment-visit-preview-link&passThrough=1)
2. Use the Einmalvergütung für Photovoltaikanlagen Dataset (you can test almost all Chart types w/ this)

3. Select Bar Chart -> Publish -> See how the tooltip is at the bottom with adjusting arrow ✅
4. Select Line Chart -> Publish -> See how the tooltip is at the bottom with adjusting arrow ✅
5. Select Area Chart -> Publish -> See how the tooltip is at the bottom with adjusting arrow ✅
6. Select Scatter Chart -> Publish -> See how the tooltip is at the bottom with adjusting arrow, but not optimal design ✅
7. Select Pie Chart -> Publish -> See how the tooltip is not placed correctly on mobile (potential separate issue), but is highlighted correctly ✅
8. Select Dual Axis Line Chart -> Publish -> See how the tooltip is at the bottom with adjusting arrow, but not optimal design ✅
9. Select Column Line Chart -> Publish -> See how the tooltip is at the bottom with adjusting arrow ✅

---

CC: @KerstinFaye , @sosiology

bprusinowski commented on app/charts/area/areas-state.tsx:
```diff
@@ -352,6 +353,9 @@ const useAreasState = (
   xScaleTimeRange.range([0, chartWidth]);
   yScale.range([chartHeight, 0]);
 
+  const theme = useTheme();
+  const isMobile = useMediaQuery(theme.breakpoints.down("sm"));
```
Nit: As this logic is used across a few different places in the application already, it would be super-nice to extract to a `useIsMobile` hook :)

noahonyejese replied:
ah cool I thought so we had something I was searching for useMobile hahaha

noahonyejese replied:
Sure  👍

noahonyejese replied:
@bprusinowski could you tell me an ideal place to put it?

bprusinowski replied:
@noahonyejese '@/utils/use-is-mobile.ts'?

bprusinowski commented on app/charts/area/areas-state.tsx:
```diff
@@ -372,18 +376,21 @@ const useAreasState = (
       });
       const xAnchor = xScale(getX(datum));
-      const yAnchor = normalize
+      const yNormalized = normalize
```
Nit: The name of the variable is only correct in "half" of the cases (when `normalize` is true) 😄 Maybe it could be renamed to `yDesktopAnchor` to make a connection with how it's used afterwards?

bprusinowski reviewed (commented):
The overall approach looks good to me, thank you @noahonyejese! There's another problem that probably was there before, that the tooltip can go out of the viewport – I wonder if we maybe can take a look at the same time?

<img width="471" alt="Screenshot 2024-11-06 at 14 42 35" src="https://github.com/user-attachments/assets/ad990ae4-3ad8-49d9-8521-c72d5e120d8f">

noahonyejese commented:
@bprusinowski yeah I think the issue you mentioned above will resolve once we implement dynamic arrows

bprusinowski commented:
Ah, but here it's not about the arrow, but that the whole tooltip can go out of viewport 👀 But maybe I missed something 👍

noahonyejese commented:
> Ah, but here it's not about the arrow, but that the whole tooltip can go out of viewport 👀 But maybe I missed something 👍

yeah technically good point its kinda both, where when we move the tooltip the arrow will be off and vise versa I will make sure it works once this is done

noahonyejese commented:
@bprusinowski I think everything works here, except for the Pie Chart, Map Chart and Scatter Plot. I will be updating the description for testing

bprusinowski commented on app/utils/use-is-mobile.ts:
```diff
@@ -0,0 +1,7 @@
+import { useMediaQuery, useTheme } from "@mui/material";
+
+export const useIsMobile = () => {
+  const theme = useTheme();
+  const isMobile = useMediaQuery(theme.breakpoints.down("sm"));
+  return isMobile;
+};
```
👍

bprusinowski commented on app/charts/shared/interaction/tooltip-box.tsx:
```diff
@@ -123,19 +127,32 @@ export const TooltipBox = ({
+  const { chartWidth } = useChartBounds(width, margins, height);
+
+  const tooltipX = isMobile
+    ? toolTipXBoundary(x!, tooltipWidth, chartWidth)
```
Nit: `tooltipXBoundary` (everywhere else we use tooltip as a single word) 👀

bprusinowski commented on app/charts/shared/interaction/tooltip-box.tsx:
```diff
@@ -158,9 +175,9 @@ export const TooltipBox = ({
-                right: triangle.right,
                 bottom: triangle.bottom,
-                left: triangle.left,
+                left: isMobile ? triangleX : triangle.left,
+                right: !isMobile ? triangle.right : undefined,
```
Nit: Would it make sense to move this logic to `mkTriangle` (could be changed to a `useTriangle` hook)? That's a nit, but I think it's not really clear why we have both `triangle.right` and `triangleX` and what is the difference between them. What do you think @noahonyejese? 🤔

noahonyejese replied:
Yeah I felt the same way about this however moving it to `mkTriangle` would force me to change the `TooltipPlacement` type to accept a new unintuitive type that kinda makes the switch case in `mkTriangle` work and I'd have to change it on a chart level rather than on a tooltip level. 

What do you think @bprusinowski ?

bprusinowski replied:
Maybe moving it to `mkTriangle` is not the best idea, but then I'd try to find a better name for these variables, or reorganize the code a bit (to clearly see a difference between e.g. `triangle.right` or `.left` and `triangleX`) 🤔

noahonyejese replied:
Yeah that's a good idea

bprusinowski replied:
Yes, I think it's much better 🎉 The only thing would be maybe `dynamicTriangleXPosition` -> `mobileTriangleXPosition`? I think that technically, the desktop one (static) is also dynamic, as it's adjusts to where the mouse is, right? If so, maybe the desktop / mobile naming convention would make sense 👍

noahonyejese replied:
Yep good idea 👍🏽

bprusinowski commented on app/charts/shared/interaction/tooltip-box.tsx:
```diff
@@ -348,3 +366,38 @@ const mkTriangle = (p: TooltipPlacement) => {
 };
+
+const toolTipXBoundary = (
+  value: number,
```
Nit: Value of what? 😄 Is this an `x` position?

bprusinowski commented on app/charts/shared/interaction/tooltip-box.tsx:
```diff
@@ -348,3 +366,38 @@ const mkTriangle = (p: TooltipPlacement) => {
+  );
+};
+
+const triangleXPos = (
+  value: number,
```
Nit: `getTriangleXPosition`, `x: number`?

bprusinowski commented on app/charts/shared/interaction/tooltip-box.tsx:
```diff
@@ -348,3 +366,38 @@ const mkTriangle = (p: TooltipPlacement) => {
+  tooltipWidth: number,
+  chartWidth: number
+): number => {
+  const condition = chartWidth - tooltipWidth / 2 + TRIANGLE_SIZE < value;
```
Nit: A `condition` for what? 😄 I think we lack an intent in the variable name here. Is this about `isOverflowing`?

bprusinowski commented on app/charts/combo/combo-line-dual-state.tsx:
```diff
@@ -165,7 +165,9 @@ const useComboLineDualState = (
       .filter(truthy);
-    const yAnchor = isMobile ? chartHeight : mean(values.map((d) => d.yPos));
+    const yAnchor = isMobile
+      ? chartHeight + 32
```
Why 32? 😄 Maybe this should be `margin.bottom`?

noahonyejese replied:
ah sorry, still had that there temp.

bprusinowski commented on app/charts/shared/interaction/tooltip-box.tsx:
```diff
@@ -383,6 +383,21 @@ const triangleXPos = (
   chartWidth: number
 ): number => {
+  if (chartWidth < tooltipWidth) {
+    const maxPosition = chartWidth + TRIANGLE_SIZE;
```
Nit: `xMax`?

bprusinowski reviewed (approved):
Thanks @noahonyejese, LGTM 👏 Just a few nit comments 👍

noahonyejese commented:
@sosiology , @KerstinFaye this is ready for testing please free to compare everything w/ the designs as well:
---
Design References:

- Scatterplot Chart: Added a ruler, but no highlighting. Position on Top if multiple Datapoints are in one line.
[Visual Design Scatterplot](https://www.figma.com/design/Hn9wvuEYLUmdJtg4KwZYUx/visualize.admin.ch---design?node-id=5850-2923&t=moajWfo6OPOxdE2X-4)

- Pie Chart: Added Highlight (border line). Position of the tooltip would stay in place.
 [Visual Design Pie Chart](https://www.figma.com/design/Hn9wvuEYLUmdJtg4KwZYUx/visualize.admin.ch---design?node-id=5847-48005&t=moajWfo6OPOxdE2X-4)

- Map Chart: Highlight is already implemented. Would keep this, but ruler not necessary since position of the Tooltip would stay in place. 
[Visual Design Map](https://www.figma.com/design/Hn9wvuEYLUmdJtg4KwZYUx/visualize.admin.ch---design?node-id=5850-3123&t=moajWfo6OPOxdE2X-4)
